### CSE4088 Introduction to Machine Learning - Homework 4

### Göksel Tokur - 150116049

### Polynomial Kernels

In [16]:
import pandas as pd
import numpy as np
from sklearn.svm import SVC

In [17]:
train_dataset = pd.read_table("http://www.amlbook.com/data/zip/features.train", header=None, names=['digit', 'intensity', 'symmetry'], sep=r'\s+')
test_dataset = pd.read_table("http://www.amlbook.com/data/zip/features.test", header=None, names=['digit', 'intensity', 'symmetry'], sep=r'\s+')


print(train_dataset.tail(5))

print(test_dataset.tail(5))

N_train = train_dataset.shape
N_test = test_dataset.shape

print("Shape of the train dataset : {}\nShape of the test dataset : {}".format(N_train, N_test))

      digit  intensity  symmetry
7286    3.0   0.296178 -2.558250
7287    3.0   0.252795 -3.110500
7288    3.0   0.393705 -2.921187
7289    0.0   0.597590 -2.258062
7290    1.0   0.276221 -0.916250
      digit  intensity  symmetry
2002    3.0   0.237842 -4.052688
2003    9.0   0.265125 -3.630500
2004    4.0   0.308857 -5.365188
2005    0.0   0.509971 -2.010000
2006    1.0   0.188977 -0.781312
Shape of the train dataset : (7291, 3)
Shape of the test dataset : (2007, 3)


In [39]:
def one_vs_all(C, Q):
    E_in_array = []
    number_of_support_vector_array = []
    for digit in range(10):
        X = train[['intensity', 'symmetry']]
        y = train['digit'] == digit
        svm = SVC(C=C, kernel='poly', degree=Q, coef0=1)
        svm.fit(X, y)
        prediction = svm.predict(X)
        E_in = np.mean(y != prediction)
        number_of_support_vector = sum(svm.n_support_)
        E_in_array.append(E_in)
        number_of_support_vector_array.append(number_of_support_vector)
        print("%d versus all E_in : %.5f\tNumber of support vectors : %d" % (digit, E_in, number_of_support_vector))
    return E_in_array, number_of_support_vector_array
        
        
        
C = 0.01
Q = 2 #degree
E_in, number_of_support_vector = one_vs_all(C=C, Q=Q)
dt = {'digit':[0, 1, 2, 3, 4, 5, 6, 7, 8, 9], 'E_in':E_in, 'Number Of Support Vector': number_of_support_vector}
dataframe = pd.DataFrame(dt)
        
    

0 versus all E_in : 0.16376	Number of support vectors : 2391
1 versus all E_in : 0.01577	Number of support vectors : 966
2 versus all E_in : 0.10026	Number of support vectors : 1525
3 versus all E_in : 0.09025	Number of support vectors : 1403
4 versus all E_in : 0.08943	Number of support vectors : 1354
5 versus all E_in : 0.07626	Number of support vectors : 1188
6 versus all E_in : 0.09107	Number of support vectors : 1400
7 versus all E_in : 0.08847	Number of support vectors : 1328
8 versus all E_in : 0.07434	Number of support vectors : 1147
9 versus all E_in : 0.08833	Number of support vectors : 1366


### Question 2

In [40]:
dataframe.loc[dataframe['E_in'].idxmax()]

digit                          0.000000
E_in                           0.163764
Number Of Support Vector    2391.000000
Name: 0, dtype: float64

As seen above, 0 versus all has the highest value (0.163764) between them.

Answer of the Question 2: [a] 0 versus all

### Question 3

In [42]:
dataframe.loc[dataframe['E_in'].idxmin()]

digit                         1.000000
E_in                          0.015773
Number Of Support Vector    966.000000
Name: 1, dtype: float64

As seen above, 1 versus all has the lowest value (0.01577) between them.

Answer of the Question 3: [a] 1 versus all

### Question 4

In [45]:
 abs(dataframe.loc[dataframe['E_in'].idxmax()]['Number Of Support Vector'] - dataframe.loc[dataframe['E_in'].idxmin()]['Number Of Support Vector'])

1425.0

Answer of the Question 4: [c] 1800

### Question 5